src/SessionManager/ProgramSettings.cc

```diff
@@ -34,6 +37,83 @@ void applyOptionalArgument(T& val, const string& argument_name, const cxxopts::P
 }
 
 ProgramSettings::ProgramSettings(int argc, char** argv) {
+    if (argc > 1) {
+        spdlog::info("Using command-line settings");
+    }
+    ApplyCommandLineSettings(argc, argv);
+
+    const fs::path user_settings_path = fs::path(getenv("HOME")) / CARTA_USER_FOLDER_PREFIX / "backend.json";
+    const fs::path system_settings_path = "/etc/carta/backend.json";
+
+    json settings;
+    if (fs::exists(system_settings_path) && !no_system_config) {
+        settings = JSONConfigSettings(system_settings_path.string());
+        system_settings_json_exists = true;
+        spdlog::info("Reading system settings from {}", system_settings_path.string());
+    }
+
+    if (fs::exists(user_settings_path) && !no_user_config) {
+        auto user_settings = JSONConfigSettings(user_settings_path.string());
+        user_settings_json_exists = true;
+        spdlog::info("Reading user settings from {}", user_settings_path.string());
+        settings.merge_patch(user_settings); // user on top of system
+    }
+
+    if (system_settings_json_exists || user_settings_json_exists) {
+        settings.merge_patch(command_line_settings); // force command-line on top of user and sytem
+        SetSettingsFromJSON(settings);
+    }
+}
+
+json ProgramSettings::JSONConfigSettings(const std::string& json_file_path) {
+    std::ifstream ifs(json_file_path, std::ifstream::in);
+    json j;
+    try {
+        j = json::parse(ifs);
+    } catch (json::exception err) {
+        spdlog::warn("Config file {} has problems, please check", json_file_path);
+        spdlog::warn(err.what());
+    }
+    return j;
+}
+
+void ProgramSettings::SetSettingsFromJSON(const json& j) {
+    for (const auto& key : int_keys_map) {
+        if (!j.contains(key.first)) {
+            continue;
+        }
+        if (!j[key.first].is_number_integer()) {
+            spdlog::warn("Config file has problems, please check key with name {}. Its current value is {}, and a number is expected",
```

Comment:
Very last comment: I would change these messages to something like:

"Problem in config file at key {}: current value is {}, and a number is expected."

It's slightly inconvenient that we can't tell the user what the file is because we've merged them, but I'm not sure if it's enough of a problem for us to re-run this function on the two files separately instead of merging them. What do you think?

Reply:
> Very last comment: I would change these messages to something like:

Feel free to add as many comments as needed. No worries.

> 
> "Problem in config file at key {}: current value is {}, and a number is expected."
> 

Thanks for the suggestion. I will change the message accordingly; also fix the number/boolean/string part such that the message is correct.

> It's slightly inconvenient that we can't tell the user what the file is because we've merged them, but I'm not sure if it's enough of a problem for us to re-run this function on the two files separately instead of merging them. What do you think?

I think it's our responsibility to give the best possible indication of what can be going wrong. I will adapt the code a little bit. It isn't a huge change. Just extending the JSONConfigSettings method to something like:
```cpp
json ProgramSettings::JSONConfigSettings(const std::string& json_file_path) {
    std::ifstream ifs(json_file_path, std::ifstream::in);
    json j;
    try {
        j = json::parse(ifs);
    } catch (json::exception err) {
        spdlog::warn("Config file {} has problems, please check", json_file_path);
        spdlog::warn(err.what());
    }
    for (const auto& key : int_keys_map) {
        if (j.contains(key.first) && !j[key.first].is_number_integer()) {
            spdlog::warn("Problem in config file {}, at key {}: current value is {}, and a number is expected.", json_file_path, key.first, j[key.first]);
        }
    }
    for (const auto& key : bool_keys_map) {
        if (j.contains(key.first) && !j[key.first].is_boolean()) {
            spdlog::warn("Problem in config file {}, at key {}: current value is {}, and a boolean is expected.", json_file_path, key.first, j[key.first]);
        }
    }
    for (const auto& key : strings_keys_map) {
        if (j.contains(key.first) && !j[key.first].is_string()) {
            spdlog::warn("Problem in config file {}, at key {}: current value is {}, and a string is expected.", json_file_path, key.first, j[key.first]);
        }
    }
    return j;
}
```
I would like to do something more elegant and get rid of those three loops and keep one with one fancy function, but I guess it's ok the way it is now.

Instead of storing valid keys elsewhere, as I attempted before with the std::set, I would rather test for their validity after merging all JSON settings again; what do you think?

Reply:
I don't love looping over the keys twice, but this probably makes more sense conceptually than fully processing each file instead of merging (especially since the last thing we're merging is commandline parameters, not a file. I do have some specific remarks, which I will add in annotations.